src/com/amazon/ion/IonCursor.java

```diff
@@ -0,0 +1,103 @@
+package com.amazon.ion;
+
+import java.io.Closeable;
+
+/**
+ * A cursor over a stream of Ion data. NOTE: this interface is not considered part of the public API.
+ */
+public interface IonCursor extends Closeable {
```

Comment:
I wonder if describing or modeling this in terms of a state machine would make it easier to grok. It _is_ an automaton at base, and the `Event` enum is more or less the current state.